base/applications/rapps/appview.cpp

```diff
@@ -970,16 +973,89 @@ CAppInfoDisplay::~CAppInfoDisplay()
 
 // **** CAppsListView ****
 
+struct CAsyncLoadIcon {
+    CAsyncLoadIcon *pNext;
+    HWND hAppsList;
+    CAppInfo *AppInfo; // Only used to find the item in the list, do not access on background thread
+    UINT TaskId;
+    bool Parse;
+    WCHAR Location[1];
+
+    void Free() { LocalFree(this); }
+    static CAsyncLoadIcon* Queue(HWND hAppsList, CAppInfo &AppInfo, bool Parse);
+    static void StartTasks();
+} *g_AsyncIconTasks = NULL;
+static UINT g_AsyncIconTaskId = 0;
+
+static DWORD CALLBACK
+AsyncLoadIconProc(LPVOID Param)
+{
+    for (CAsyncLoadIcon *task = (CAsyncLoadIcon*)Param, *old; task; old->Free())
+    {
+        if (task->TaskId == g_AsyncIconTaskId)
+        {
+            HICON hIcon;
+            if (!task->Parse)
+                hIcon = (HICON)LoadImageW(NULL, task->Location, IMAGE_ICON, 0, 0, LR_LOADFROMFILE | LR_DEFAULTSIZE);
+            else if (!ExtractIconExW(task->Location, PathParseIconLocationW(task->Location), &hIcon, NULL, 1))
+                hIcon = NULL;
+
+            if (hIcon)
+            {
+                SendMessageW(task->hAppsList, WM_RAPPSLIST_ASYNCICON, (WPARAM)hIcon, (LPARAM)task);
+                DestroyIcon(hIcon);
+            }
+        }
+        old = task;
+        task = task->pNext;
+    }
+    return 0;
+}
+
+CAsyncLoadIcon*
+CAsyncLoadIcon::Queue(HWND hAppsList, CAppInfo &AppInfo, bool Parse)
+{
+    ATLASSERT(GetCurrentThreadId() && GetWindowThreadProcessId(hAppsList, NULL));
+    CStringW szIconPath;
+    if (!AppInfo.RetrieveIcon(szIconPath))
+        return NULL;
+    SIZE_T cch = szIconPath.GetLength() + 1, cbstr = cch * sizeof(WCHAR);
+    CAsyncLoadIcon *task = (CAsyncLoadIcon*)LocalAlloc(LMEM_FIXED, sizeof(CAsyncLoadIcon) + cbstr);
+    if (!task)
+        return NULL;
+    task->hAppsList = hAppsList;
+    task->AppInfo = &AppInfo;
+    task->TaskId = g_AsyncIconTaskId;
+    task->Parse = Parse;
+    CopyMemory(task->Location, szIconPath.GetBuffer(), cbstr);
+    szIconPath.ReleaseBuffer();
+    task->pNext = g_AsyncIconTasks;
+    g_AsyncIconTasks = task;
```

Comment:
Can this `CAsyncLoadIcon::Queue()` helper be called concurrently?
If so, this `g_AsyncIconTasks` should probably be `InterlockedExchangePointer` 'ed...
Or maybe, use a [SLIST](https://learn.microsoft.com/en-us/windows/win32/api/interlockedapi/nf-interlockedapi-interlockedpushentryslist) or something where queuing can be done atomically.

Reply:
It cannot, all of this is happening on the UI thread. And the assert is there to make sure it stays that way.